openedx_learning/apps/authoring/linking/models.py

```diff
@@ -0,0 +1,132 @@
+"""
+Contains models to store upstream-downstream links between publishable entities.
+"""
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from openedx_learning.lib.fields import (
+    case_insensitive_char_field,
+    case_sensitive_char_field,
+    immutable_uuid_field,
+    key_field,
+    manual_date_time_field,
+)
+
+from ..publishing.models import PublishableEntity
+
+__all__ = [
+    "PublishableEntityLink",
+    "CourseLinksStatus",
+    "CourseLinksStatusChoices"
+]
+
+
+class PublishableEntityLink(models.Model):
+    """
+    This represents link between any two publishable entities or link between publishable entity and a course
+    xblock. It helps in tracking relationship between xblocks imported from libraries and used in different courses.
+    """
+    uuid = immutable_uuid_field()
+    upstream_block = models.ForeignKey(
+        PublishableEntity,
+        on_delete=models.SET_NULL,
+        related_name="links",
+        null=True,
+        blank=True,
+    )
+    upstream_usage_key = key_field(
+        help_text=_(
+            "Upstream block usage key, this value cannot be null"
+            " and useful to track upstream library blocks that do not exist yet"
+        )
+    )
+    upstream_context_key = case_sensitive_char_field(
```

Comment:
Why isn't this a `key_field` like the `downstream_context_key` and `CourseLinksStatus.context_key`?

Reply:
Initially I was planning to keep the field nullable which is why I did not use `key_field` but after giving it a bit more thought, used `key_field` and made sure that field is always added.